osquery/tables/smart/smart_drives.cpp

```diff
@@ -0,0 +1,221 @@
+#include "osquery/tables/smart/smart_drives.h"
+
+namespace osquery {
+namespace tables {
+
+static inline std::ostream& operator<<(
+    std::ostream& stream, const std::vector<hardwareDriver>& devs) {
+  for (const auto& dev : devs) {
+    stream << " " << dev.driver;
+  }
+
+  return stream;
+}
+
+/// Static map of supported controller types.
+static const std::map<std::string, hardwareDriver>
+    kSmartExplicitDriverToDevice = {
+        {"megaraid_sas", hardwareDriver{"megaraid,", 127}},
+        {"hpsa", hardwareDriver{"cciss,", 14}},
+};
+
+/// Gets all devices of class 'Mass storage controller'.
+void getStorageCtlerClassDrivers(std::vector<std::string>& results) {
+  auto devices = SQL::selectAllFrom("pci_devices");
+  for (const auto& device : devices) {
+    auto pciClass = device.find("pci_class");
+    auto driver = device.find("driver");
+    if (pciClass != device.end() &&
+        pciClass->second == "Mass storage controller" &&
+        driver != device.end()) {
+      auto driverName = driver->second;
+      auto i = std::lower_bound(results.begin(), results.end(), driverName);
+      if (i == results.end() || driverName < *i) {
+        results.insert(i, std::move(driverName));
+      }
+    }
+  }
+}
+
+/// Gets supported hardwareDriver from system storage controllers.
+static inline void getSmartCtlDeviceType(
+    const std::vector<std::string>& storageDrivers,
+    std::vector<hardwareDriver>& types) {
+  for (const auto& driver : storageDrivers) {
+    auto hwDriver = kSmartExplicitDriverToDevice.find(driver);
+    if (hwDriver != kSmartExplicitDriverToDevice.end()) {
+      types.push_back(hwDriver->second);
+    }
+  }
+}
+
+/// Utility function for traversing system devices.
+void walkBlkDevices(std::function<void(const std::string& devname,
+                                       hardwareDriver* type)> handleDevF) {
+  if (getuid() != 0 || geteuid() != 0) {
+    LOG(WARNING) << "Need root access for smart information";
+    return;
+  }
+
+  std::vector<std::string> storageDrivers;
+  getStorageCtlerClassDrivers(storageDrivers);
+
+  std::vector<hardwareDriver> types;
+  getSmartCtlDeviceType(storageDrivers, types);
+
+  if (types.size() > 1) {
+    LOG(WARNING) << "Found more than 1 hardware storage controller:" << types
+                 << "; only handling the first";
+    return;
+  }
+
+  hardwareDriver* type = nullptr;
+  if (types.size() == 1) {
+    type = &(types[0]);
+  }
+
+  auto blkDevices = SQL::selectAllFrom("block_devices");
+  for (const auto& device : blkDevices) {
+    auto size = device.find("size");
+    auto name = device.find("name");
+    if (size != device.end() && size->second != "0" && size->second != "" &&
+        name != device.end()) {
+      handleDevF(name->second, type);
+    }
+  }
+}
+
+void querySmartDevices(
+    libsmartctl::ClientInterface& c,
+    std::function<
+        void(std::function<void(const std::string&, hardwareDriver*)>)> walkF,
```

Comment:
nit: similarly here, no need to keep it short, let's use something more descriptive, perhaps `walk_func` or `walk_function`?